chain/pool/src/lib.rs

```diff
@@ -49,10 +53,16 @@ impl TransactionPool {
     }
 
     /// Insert a signed transaction into the pool that passed validation.
-    pub fn insert_transaction(&mut self, signed_transaction: SignedTransaction) -> bool {
+    pub fn insert_transaction(
+        &mut self,
+        signed_transaction: SignedTransaction,
+    ) -> Result<bool, Error> {
+        if self.unique_transactions.len() >= self.max_unique_transactions {
+            return Err(Error::TransactionPoolFull);
+        }
         if !self.unique_transactions.insert(signed_transaction.get_hash()) {
             // The hash of this transaction was already seen, skip it.
-            return false;
+            return Ok(false);
         }
```

Comment:
Those two checks should be switched. Or rather, if we reached the size limit we should check if the transaction already exists and then return `Ok(false)`.